### PR TITLE
feat: 3rd rewrite, drop handlerSync

### PR DESCRIPTION
**Pull request by tunnckoCore**
related to #1 

- 3rd rewrite
- drop handlerSync - handle async through returned promises
- introduce `.options` - tho, need fixes to actually make it work
- working latest files: `finalizing.ts`, `utils.ts` and `zagora-v3-types.ts`
- doesn't support `errorsFirst` yet, and maybe it should be `errorHelpersFirst`

maybe drop the `.options`... it's weird, especially coupled with `z.tuple` args.. like, if both are defined, and there is an optional/default set as the second item in tuple schema then what? It's best if `.input(z.string()).options(z.object({ foo: z.string() }))` but.. the `z.tuple` works perfect and handles all cases so it's just needed to do `z.tuple([z.string(), z.object({ foo: z.string().default('foo') }).optional()])`

## Summary by CodeRabbit

* **New Features**
  * Simplified builder API, new typed ZagoraError and error-helper utilities; added example router/procedures and domain schemas for demos.

* **Bug Fixes / Behavior**
  * Unified validation and error handling (now thrown/reported consistently); updated error shapes, tuple/default handling, and mixed-schema behavior.

* **Tests**
  * Large expansion of tests covering sync/async flows, tuples/arrays, mixed schemas, and many error scenarios.

* **Documentation**
  * README refreshed with usage, API summary, examples and notes on error/default behavior.

* **Chores**
  * Tooling, formatter/linter configs, package scripts and CI/release workflows updated; example file removed.

### CODE REVIEW

**Comment by tunnckoCore** [comment]
there is a known bug with valibot tuples - when an item in a tuple is optional TypeScript fails to resolve properly the overloaded input arguments of created handler.

check out the below test and the comments 

```ts
test("Valibot tuple with optional and default value", async () => {
  const SpeedSchema = v.picklist(["slow", "normal", "fast"]);

  const hello = zagora()
    .input(v.tuple([SpeedSchema, v.optional(v.number(), 123)]))
    .output(
      v.object({
        foo: v.pipe(v.string(), v.minLength(1)),
      })
    )
    .handler(async (speed, retry) => {
      return { foo: `${speed}-${retry}` };
    });

  // TODO: expected error for valibot; for zod it works; it does work for per-arg type validation
  // const [resHello, errHello] = await hello("fast"); // would signal incorrectly at `hello` for missing second arg
  // const [resHello, errHello] = await hello("fast", "sasa"); // would signal `"sasa"` that it expects number
  // const [resHello, errHello] = await hello("fast", 123); // would not type error, all args are fine and provided

  // @ts-expect-error expected for valibot
  const [resHello, errHello] = await hello("fast");

  expect(errHello).toBe(null);
  expect(resHello).toEqual({ foo: "fast-123" });
});
```

when we hover over the `hello`  in `const hello` we get

```ts
(args_0: "slow" | "normal" | "fast", args_1: number | undefined) => 
	Promise<
		Result<null, ZagoraError, false> | 
		Result<{   foo: string; }, null, false>
	>
```

while if we use the exact same example but use Zod schemas instead like so...


```ts
test("Zod tuple with default values - basic case", async () => {
  const SpeedSchema = z.enum(["slow", "normal", "fast"]);

  const hello = zagora()
    .input(z.tuple([SpeedSchema, z.number().default(123)]))
    .output(
      z.object({
        foo: z.string().min(1),
      })
    )
    .handler((speed, retry) => {
      return { foo: `${speed}-${retry}` };
    });

  // no type errors since typescript properly resolve optionals
  const [resHello, errHello] = hello("fast"); 

  expect(errHello).toBe(null);
  expect(resHello).toEqual({ foo: "fast-123" });
});
```

 putting aside the messy merging and etc.. it does properly get what's optional and what's not

```ts
const hello: ((args_0: "slow" | "normal" | "fast", args_1?: number | undefined) => Promise<Result<null, ZagoraError, false> | Result<{
    foo: string;
}, null, false>>) & ((args_0: "slow" | "normal" | "fast", args_1?: number | undefined) => Promise<Result<null, ZagoraError, false> | Result<{
    foo: string;
}, null, false>>) & ForwardWithHandler<((args_0: "slow" | "normal" | "fast", args_1?: number | undefined) => Promise<Result<null, ZagoraError, false> | Result<{
    foo: string;
}, null, false>>) & ((args_0: "slow" | "normal" | "fast", args_1?: number | undefined) => Promise<Result<null, ZagoraError, false> | Result<{
    foo: string;
}, null, false>>)>
```

Check out the `edge-cases.test.ts` and `zagora-v3-types.ts`. We are currently using `finalizing.ts` instead of the `index.ts` as main.

**Comment by tunnckoCore** [comment]
Another way would be to update the OverloadedByPrefix to

```ts
export type OverloadedByPrefixes<T extends any[], R> = UnionToIntersection<
  ValuePrefixes<T> extends infer P
    ? P extends any[]
      ? P extends []
        ? AllOptional<T> extends true
          ? (...args: P) => R
          : never
        : (...args: P) => R
      : never
    : never
>;
```

eg. without the `AllOptional<T>` check which would report on per argument basis which is cool but the signature when you hover on `hello` would show only the require ones at first :D

**Comment by tunnckoCore** [comment]
calling the typescript gods @mattpocock @colinhacks @fabian-hiller for debugging this.

i came up with a partial solution of OverloadedByPrefixes, but still.. i don't get how or why different libraries, but StandardSchema compatible, give different results, eg. typescript resolve to different paths in generics ternaries.

Everything runtime works, it's just a type system problem.

Would appreciate help to get to the bottom of this.

**Comment by fabian-hiller** [comment]
Hey, I am very sorry. I don't have a lot of time right now. 😐 But your library looks interesting. Could you solve the problem you have?

**Comment by tunnckoCore** [comment]
@fabian-hiller @colinhacks

yeah, Haiku figured it out:

> **The Problem:** Valibot represents optional tuple elements as `[string, number | undefined]` (union with undefined), while Zod represents them as `[string, number?]` (optional property). This causes TypeScript to require both arguments at the callsite even though the second has a default value.
> 
> **The Solution:** I added a `MakeUndefinedOptional` helper type that detects when tuple elements have `undefined` in their type and converts them to optional parameters `[string, number?]`. This only activates for Valibot cases (when undefined is present) and doesn't affect Zod, so it's safe and doesn't break existing code.
> 
> **Result:** Now both `hello("fast")` and `hello("slow", 5)` work correctly with Valibot optional tuples with defaults, matching Zod's behavior.

File: basic-example.ts [hunk 1]
```diff
@@ -1,23 +1,62 @@
...
+// const [resHello, errHello] = await hello("fast"); // would signal incorrectly at `hello` for missing second arg
+// const [resHello, errHello] = await hello("fast", "sasa"); // would signal `"sasa"` that it expects number
+// const [resHello, errHello] = await hello("fast", 123); // would not type error, all args are fine and provided
+
+// @ts-expect-error expected for valibot
 const [resHello, errHello] = await hello("fast");
 
-console.log({ resHello, errHello });
+expect(errHello).toBe(null);
+expect(resHello).toEqual({ foo: "fast-123" });
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply basic-example.ts]: ignore, it's example file i just play with

File: src/utils.ts [hunk 2]
```diff
@@ -1,50 +1,317 @@
...
+  if (obj === "[object AsyncFunction]") {
+    return true;
+  }
+
+  try {
+    const result = fn();
+    return result instanceof Promise;
+  } catch (_err: unknown) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply src/utils.ts:41]: errmm.. no, the whole point is to detect if it's a promise or not, and a try catch is a must to not crash.

File: old_src/zagora-v3.ts [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+
+      if (resp instanceof Promise) {
+        return resp.then((x) => createResult(x.data, x.error, x.isDefined));
+      }
+
+      return createResult(resp.data, resp.error, resp.isDefined);
+    };
+    const forward = forwardImpl as unknown as ForwardType;
+
+    return forward;
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply old_src/zagora-v3.ts]: no thanks

File: README.md [hunk 4]
```diff
@@ -126,13 +137,13 @@ const uppercaseString = zagora()
...
 		if (input === "network") {
-			return err.network({
+			throw err.network({
 				message: "Network failed",
 				statusCode: 500,
 			});
 		}
 		if (input === "validation") {
-			return err.validation({
+			throw err.validation({
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply README.md]: fixed

File: test/edge-cases.test.ts [hunk 5]
```diff
@@ -49,15 +99,15 @@ test("Zod tuple with default values - both args provided", async () => {
...
+        .strict(),
     )
     .handler(async (speed, retry) => {
       return { foo: `${speed}-${retry}`, bar: "barry" };
     });
 
   const [resHello2, errHello2] = await hello2("slow", 456);
 
-  expect(errHello2?.reason).toContain("Failure caused by validation");
+  expect(errHello2?.reason).toContain("utput validation failed");
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply test/edge-cases.test.ts:110]: not, that's intentional!

File: README.md [hunk 6]
```diff
@@ -162,28 +173,99 @@ if (error && isDefined) {
...
+  })
+```
+
+So, be careful when using `.default` or `.optional` method on schemas. While both cases are basically saying the same thing, they are different things on type-system level.
+
+### Note on errors & no input schema
+
+It's important to note, that when you don't want or need to have an input schema, but want typed errors, then you should expect the `errors` object with typed error helpers to be on the second argument, not the first, even though there's no inputs.
+
+For example, the following is valid, and it's kinda known limitation of the current type system, but it's fine since it's probably a rare edge case.
```
[the remark this answers is not in the document]
Comment (tunnckoCore) [reply README.md]: done
